h2o-core/src/main/java/water/Weaver.java

```diff
@@ -36,166 +32,49 @@
+      throw new RuntimeException(e);
+    }
   }
 
   // The name conversion from a Iced subclass to an Icer subclass.
   private static String implClazzName( String name ) {
     return name + "$Icer";
   }
 
-//  private static boolean hasWovenJSONFields( CtClass cc ) throws NotFoundException {
-//    if( !cc.subtypeOf(_freezable) &&
-//        !cc.subtypeOf(_serialize) ) return false; // Cannot serialize in any case
-//    // Iced & H2O$CountedCompleters are interesting oddballs: they have a short
-//    // typeid that is desired field for Freezable-style serialization but not for
-//    // JSON-style.  The field is fairly expensively filled in the first time any
-//    // given object is serialized and used in all subsequent fast serializations.
-//    // However, the value is not valid outside *this* execution of the cluster,
-//    // and should not be persisted via e.g. saving the JSON and restoring from
-//    // it later.
-//    if( cc.equals(_iced) ||
-//        cc.equals(_h2cc) ) return false;
-//    if( hasWovenJSONFields(cc.getSuperclass()) ) return true;
-//    for( CtField ctf : cc.getDeclaredFields() ) {
-//      int mods = ctf.getModifiers();
-//      if( !javassist.Modifier.isTransient(mods) && !javassist.Modifier.isStatic(mods) ) return true;
-//    }
-//    return false;
-//  }
-
-  /**
-   * Load/Reload classes defined at runtime.
-   *
-   * Loading classes at runtime is a matter of simply injecting the
-   * new code into the {@link ClassPool}, and then {@link Weaver#javassistLoadClass(int, Class)}
-   * resolves the generation of (de)serializers. In order to reload classes, though,
-   * each dynamically loaded class must have its very own {@link ClassLoader}, and all
-   * previous {@link Icer}s must be removed. In order to maintain cluster-wide coherency
-   * about which classes are loaded, the {@link TypeMap} is likewise updated whenever a class
-   * is reloaded.
-   *
-   * In order to successfully load classes at runtime (for example a subclass of {@link MRTask}),
-   * each node takes the bytecode and class name and puts a new {@link ByteArrayClassPath} onto
-   * {@link Weaver#_pool}'s classpath. Since there is no mechanism for retrieving these
-   * {@link ClassPath} instances later, they are stored in {@link Weaver#CLASSPATHS} so that
-   * reload events can remove the old paths. Similarly, {@link Weaver#CLASSLOADERS} holds on
-   * to the loaders of dynamically created classes so that classes can be reloaded and old
-   * {@link ClassLoader} instances pruned.
-   *
-   * Finally, in order to {@link Weaver#genDelegate(int, Class)} during a class reload, then the
-   * previous {@link Icer} must be {@link CtClass#detach}ed. In addition, the {@link TypeMap#goForGold(int)}
-   * must turn a null for the {@link Icer}.
-   *
-   * @param name class name
-   * @param b bytecode
-   */
-  public static void loadDynamic(final String name, final byte[] b) {
```

Comment:
unused? What was the use-case for this? Plugins/extensions? Custom metrics? The unit test doesn't help…

Reply:
This basically lets you modify Iced classes at runtime and re-generate Icers.

To be honest I don't know if this was ever used and how but it definitely is not used for extension points and by 3rd parties. Custom extensions are put on classpath before H2O boots up and classloading happens the regular way.

To my knowledge very safe to remove.